dbr quant: extend qconfig_dict support to functions, part 1

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #69766 dbr quant: extend qconfig_dict support to functionals, part 2
* **#69758 dbr quant: extend qconfig_dict support to functions, part 1**
* #69726 dbr quant: make fqn during prepare op hook required
* #69720 fx quant: move _parent_name to common utils
* #69719 dbr quant: initial support of qconfig_dict for modules
* #69636 fx quant: refactor qconfig_dict utils to separate file

Summary:

Extends DBR quant `qconfig_dict['object_type']` support to function types,
with the restriction that a parent module must have a qconfig.

A future PR will remove the restriction above (it is due to some technical
debt), to keep PR sizes small.

Test plan:

```
python test/test_quantization.py TestQuantizeDBR
```

Differential Revision: [D33020217](https://our.internmc.facebook.com/intern/diff/D33020217)

vkuzo committed:
dbr quant: extend qconfig_dict support to functions

[ghstack-poisoned]

vkuzo committed:
Update on "dbr quant: extend qconfig_dict support to functions, part 1"


Summary:

Extends DBR quant `qconfig_dict['object_type']` support to function types,
with the restriction that a parent module must have a qconfig.

A future PR will remove the restriction above (it is due to some technical
debt), to keep PR sizes small.

Test plan:

```
python test/test_quantization.py TestQuantizeDBR
```

[ghstack-poisoned]

vkuzo committed:
Update on "dbr quant: extend qconfig_dict support to functions, part 1"


Summary:

Extends DBR quant `qconfig_dict['object_type']` support to function types,
with the restriction that a parent module must have a qconfig.

A future PR will remove the restriction above (it is due to some technical
debt), to keep PR sizes small.

Test plan:

```
python test/test_quantization.py TestQuantizeDBR
```

Differential Revision: [D33020217](https://our.internmc.facebook.com/intern/diff/D33020217)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33020217).

vkuzo committed:
Update on "dbr quant: extend qconfig_dict support to functions, part 1"


Summary:

Extends DBR quant `qconfig_dict['object_type']` support to function types,
with the restriction that a parent module must have a qconfig.

A future PR will remove the restriction above (it is due to some technical
debt), to keep PR sizes small.

Test plan:

```
python test/test_quantization.py TestQuantizeDBR
```

Differential Revision: [D33020217](https://our.internmc.facebook.com/intern/diff/D33020217)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33020217).

vkuzo committed:
Update on "dbr quant: extend qconfig_dict support to functions, part 1"


Summary:

Extends DBR quant `qconfig_dict['object_type']` support to function types,
with the restriction that a parent module must have a qconfig.

A future PR will remove the restriction above (it is due to some technical
debt), to keep PR sizes small.

Test plan:

```
python test/test_quantization.py TestQuantizeDBR
```

Differential Revision: [D33020217](https://our.internmc.facebook.com/intern/diff/D33020217)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33020217).

vkuzo committed:
Update on "dbr quant: extend qconfig_dict support to functions, part 1"


Summary:

Extends DBR quant `qconfig_dict['object_type']` support to function types,
with the restriction that a parent module must have a qconfig.

A future PR will remove the restriction above (it is due to some technical
debt), to keep PR sizes small.

Test plan:

```
python test/test_quantization.py TestQuantizeDBR
```

Differential Revision: [D33020217](https://our.internmc.facebook.com/intern/diff/D33020217)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33020217).

vkuzo committed:
Update on "dbr quant: extend qconfig_dict support to functions, part 1"


Summary:

Extends DBR quant `qconfig_dict['object_type']` support to function types,
with the restriction that a parent module must have a qconfig.

A future PR will remove the restriction above (it is due to some technical
debt), to keep PR sizes small.

Test plan:

```
python test/test_quantization.py TestQuantizeDBR
```

Differential Revision: [D33020217](https://our.internmc.facebook.com/intern/diff/D33020217)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33020217).

vkuzo committed:
Update on "dbr quant: extend qconfig_dict support to functions, part 1"


Summary:

Extends DBR quant `qconfig_dict['object_type']` support to function types,
with the restriction that a parent module must have a qconfig.

A future PR will remove the restriction above (it is due to some technical
debt), to keep PR sizes small.

Test plan:

```
python test/test_quantization.py TestQuantizeDBR
```

Differential Revision: [D33020217](https://our.internmc.facebook.com/intern/diff/D33020217)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33020217).

vkuzo committed:
Update on "dbr quant: extend qconfig_dict support to functions, part 1"


Summary:

Extends DBR quant `qconfig_dict['object_type']` support to function types,
with the restriction that a parent module must have a qconfig.

A future PR will remove the restriction above (it is due to some technical
debt), to keep PR sizes small.

Test plan:

```
python test/test_quantization.py TestQuantizeDBR
```

Differential Revision: [D33020217](https://our.internmc.facebook.com/intern/diff/D33020217)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33020217).

vkuzo committed:
Update on "dbr quant: extend qconfig_dict support to functions, part 1"


Summary:

Extends DBR quant `qconfig_dict['object_type']` support to function types,
with the restriction that a parent module must have a qconfig.

A future PR will remove the restriction above (it is due to some technical
debt), to keep PR sizes small.

Test plan:

```
python test/test_quantization.py TestQuantizeDBR
```

Differential Revision: [D33020217](https://our.internmc.facebook.com/intern/diff/D33020217)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33020217).

jerryzh168 commented on test/quantization/dbr/test_quantize_dbr.py:
```diff
@@ -1025,6 +1027,40 @@ def test_qconfig_dict_object_type_module(self):
+        mp(*example_args)
+        mq = _quantize_dbr.convert(mp)
+        mq(*example_args)
+        rewritten = mq.rewrite_for_scripting()
```
just curious, is there a way to check what ops are run in dbr quantized models? or do we plan to build some utilities to collect that information?

vkuzo replied:
Currently the developer can turn on logging (by hardcoding a flag in `auto_trace.py`) to inspect it in real time.  I think the cleanest way to do it in tests is to encode the transforms into the graph and check the graph, this can be done with `torch.jit.trace` or `torch.jit.script`.  The current code uses the FX bridge for `torch.jit.script`.

I'm flexible if we want to do something else here, lmk.

jerryzh168 commented on torch/ao/quantization/_dbr/qconfig_dict_utils.py:
```diff
@@ -0,0 +1,27 @@
+
+TYPE_TO_REPLACEMENT_TYPE = {
+    torch.add: torch.Tensor.add,
+    torch.Tensor.add_: torch.Tensor.add,
```
do we have `torch.Tensor.add` as well?

vkuzo replied:
The user code can have any variant of add, but `torch.Tensor.add` is what we see inside the `__torch_function__` dispatch.  DBR quant only sees the types inside of __torch_function__, so it does not have any knowledge of which variant was in the user's code.

jerryzh168 commented on test/quantization/dbr/test_quantize_dbr.py:
```diff
@@ -1025,6 +1027,40 @@ def test_qconfig_dict_object_type_module(self):
+        # and individual functions having qconfigs
+        qconfig_dict = {
+            '': torch.quantization.default_qconfig,
+            'object_type': [
```
in fx, 'object_type' can be module_type, torch function/torch ops, and methods (call_method node in fx), do we plan to support modules and call_methods as well? currently we use strings for call_method, e.g. "add", which probably means `torch.Tensor.add` here, maybe we want to standardize this api as well. I feel it might be better to use `torch.Tensor.add` as the standard format, and have a map from string to it in fx, what do you think?

jerryzh168 replied:
looks like the others might also be supported already, can we add two more tests, one for module, one for method (torch.Tensor.add)?

vkuzo replied:
sure, the module tests are added in https://github.com/pytorch/pytorch/pull/69719 (earlier in this stack), we can add the method test to this PR

vkuzo replied:
let me cover methods in a separate PR

vkuzo replied:
> I feel it might be better to use torch.Tensor.add as the standard format, and have a map from string to it in fx, what do you think?

yes, after looking into methods in DBR, I agree with that.

jerryzh168 commented on test/quantization/dbr/test_quantize_dbr.py:
```diff
@@ -1025,6 +1027,40 @@ def test_qconfig_dict_object_type_module(self):
+        """
+        class M(nn.Module):
+            def forward(self, x):
+                x = x + x
```
actually, isn't this operator.add?

vkuzo replied:
DBR quant will see this after all the PyTorch transformations, as `torch.Tensor.add`, so the knowledge that the user code had `operator.add` is not something that is visible to DBR quant.  I was planning to add test coverage for all the types of add in a future PR, this PR is just to test the path for qconfig_dict with a single thing.

vkuzo committed:
Update on "dbr quant: extend qconfig_dict support to functions, part 1"


Summary:

Extends DBR quant `qconfig_dict['object_type']` support to function types,
with the restriction that a parent module must have a qconfig.

A future PR will remove the restriction above (it is due to some technical
debt), to keep PR sizes small.

Test plan:

```
python test/test_quantization.py TestQuantizeDBR
```

Differential Revision: [D33020217](https://our.internmc.facebook.com/intern/diff/D33020217)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33020217).

vkuzo committed:
Update on "dbr quant: extend qconfig_dict support to functions, part 1"


Summary:

Extends DBR quant `qconfig_dict['object_type']` support to function types,
with the restriction that a parent module must have a qconfig.

A future PR will remove the restriction above (it is due to some technical
debt), to keep PR sizes small.

Test plan:

```
python test/test_quantization.py TestQuantizeDBR
```

Differential Revision: [D33020217](https://our.internmc.facebook.com/intern/diff/D33020217)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33020217).
